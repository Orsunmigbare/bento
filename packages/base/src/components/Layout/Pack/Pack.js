import React from "react"
import PropTypes from "prop-types"
import styled from "styled-components"

const Pack = ({ children, className, bottom, middle, wrap }) => (
  <PackUl className={className} bottom={bottom} middle={middle} wrap={wrap}>
    {children}
  </PackUl>
)

const position = (bottom, middle) => {
  if (bottom === true) return "flex-end"
  else if (middle === true) return "center"
}

const PackUl = styled.ul`
  width: 100%;
  margin-left: 0;
  display: flex;
  justify-content: space-between;
  align-items: ${p => position(p.bottom, p.middle)};
  flex-wrap: ${p => p.wrap && "wrap"};
`

Pack.defaultProps = {
  className: "",
  bottom: false,
  middle: false,
  wrap: false
}

Pack.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element),
    PropTypes.element
  ]).isRequired,
  className: PropTypes.string,
  bottom: PropTypes.bool,
  middle: PropTypes.bool,
  wrap: PropTypes.bool
}

export default Pack
