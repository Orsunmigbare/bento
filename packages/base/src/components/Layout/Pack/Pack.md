The pack object simply causes any number of elements pack up horizontally to automatically fill an equal, fluid width of their parent.

### Top-aligned

```jsx
<Pack>
  <PackItem>Lån</PackItem>
  <PackItem>Kreditt</PackItem>
  <PackItem>Sparing</PackItem>
  <PackItem>Om denne banken</PackItem>
  <PackItem>Kontakt</PackItem>
</Pack>
```

### Middle-aligned

```jsx
<Pack middle>
  <PackItem>Lån</PackItem>
  <PackItem>Kreditt</PackItem>
  <PackItem>Sparing</PackItem>
  <PackItem>Om denne banken</PackItem>
  <PackItem>Kontakt</PackItem>
</Pack>
```

### Bottom-aligned

```jsx
<Pack bottom>
  <PackItem>Lån</PackItem>
  <PackItem>Kreditt</PackItem>
  <PackItem>Sparing</PackItem>
  <PackItem>Om denne banken</PackItem>
  <PackItem>Kontakt</PackItem>
</Pack>
```
