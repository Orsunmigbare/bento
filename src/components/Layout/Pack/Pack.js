import React from "react"
import PropTypes from "prop-types"
import styled from "styled-components"
import { applyVariants } from "../../../theming"
import themeProps from "./Pack.themeProps"

/**
 * The pack object simply causes any number of elements pack up horizontally to automatically fill an equal, fluid width of their parent.
 */
const Pack = ({ children, className, bottom, middle, noWrap, justify }) => (
  <PackUl
    className={className}
    bottom={bottom}
    middle={middle}
    noWrap={noWrap}
    justify={justify}
  >
    {children}
  </PackUl>
)

const position = (bottom, middle) => {
  if (bottom === true && middle === true)
    console.warn("Pack cannot be middle and bottom at the same time")
  else if (bottom === true) return "flex-end"
  else if (middle === true) return "center"
}

export const PACK = "PACK"
const PackUl = styled.ul`
  width: 100%;
  margin-left: 0;
  display: flex;
  justify-content: ${p => p.justify};
  align-items: ${p => position(p.bottom, p.middle)};
  flex-wrap: ${p => !p.noWrap && "wrap"};
  ${applyVariants(themeProps.list)};
`

Pack.defaultProps = {
  className: "",
  bottom: false,
  middle: false,
  noWrap: false,
  justify: "space-between"
}

Pack.propTypes = {
  /**
   * The component require a child and it can be an array of elements or one element.
   */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element),
    PropTypes.element
  ]).isRequired,
  /**
   * The component can be assigned a class.
   */
  className: PropTypes.string,
  /**
   * The position of the Pack
   */
  bottom: PropTypes.bool,
  /**
   * The position of the Pack, a boolean
   */
  middle: PropTypes.bool,
  /**
   * The position of the Pack, a boolean
   */
  noWrap: PropTypes.bool,
  /**
   * The prop accepts a string, e.g. justify="flex-start".
   */
  justify: PropTypes.string
}
Pack.themeProps = themeProps

export default Pack
