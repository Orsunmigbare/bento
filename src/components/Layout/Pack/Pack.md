The pack object simply causes any number of elements pack up horizontally to automatically fill an equal, fluid width of their parent.

### Top-aligned

```jsx
<Pack>
  <PackItem>Lån</PackItem>
  <PackItem>Kreditt</PackItem>
  <PackItem>Sparing</PackItem>
  <PackItem>Om denne banken</PackItem>
  <PackItem>Kontakt</PackItem>
</Pack>
```

### Middle-aligned and noWrapped

```jsx
<Pack middle noWrap>
  <PackItem>Lån</PackItem>
  <PackItem>Kreditt</PackItem>
  <PackItem>Sparing</PackItem>
  <PackItem>Om denne banken</PackItem>
  <PackItem>Kontakt</PackItem>
</Pack>
```

### Bottom-aligned

Bottom, middle and justify flex-start

```jsx
<Pack bottom middle justify={"flex-start"}>
  <PackItem>Lån</PackItem>
  <PackItem>Kreditt</PackItem>
  <PackItem>Sparing</PackItem>
  <PackItem>Om denne banken</PackItem>
  <PackItem>Kontakt</PackItem>
</Pack>
```
