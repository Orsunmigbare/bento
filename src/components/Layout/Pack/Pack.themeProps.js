import { themePropTypes } from "../../../theming"

const themeProps = {
  list: {
    name: "PACK",
    description: "",
    type: themePropTypes.style
  }
}

export default themeProps
