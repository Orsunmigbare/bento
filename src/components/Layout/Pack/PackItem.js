import React from "react"
import PropTypes from "prop-types"
import styled from "styled-components"
import { applyVariants, spacing, commonPropTypes } from "../../../theming"

const PackItem = ({ className, children, ...rest }) => (
  <ListItem className={className} {...rest}>
    {children}
  </ListItem>
)

export const PACK_ITEM = "PACK_ITEM"
const ListItem = styled.li`
  display: inline-block;
  padding: ${spacing.small};
  ${applyVariants(PACK_ITEM)};
`

PackItem.defaultProps = {
  className: ""
}

PackItem.propTypes = {
  /**
   * The component require a child.
   */
  children: commonPropTypes.children.isRequired,
  /**
   * The component can be assigned a class.
   */
  className: PropTypes.string
}

export default PackItem
