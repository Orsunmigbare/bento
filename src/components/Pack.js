import React from "react"
import styled from "styled-components"
import { spacing } from "@staccx/theme"

const Pack = ({ children }) => (
  <Wrap>
    <PackElement>{children}</PackElement>
  </Wrap>
)

const Wrap = styled.div`
  overflow: hidden;
`
const PackElement = styled.div`
  display: flex;
  flex-wrap: wrap;
  margin: -${spacing.tiny};
  & > * {
    margin: ${spacing.tiny};
  }
`

export default Pack
