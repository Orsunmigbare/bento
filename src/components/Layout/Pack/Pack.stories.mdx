import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import Pack from './Pack';
import PackItem from './PackItem';

<Meta title="Components|Layout/Pack and PackItem" component={Pack, PackItem}/>

# Pack and PackItem
The pack object simply causes any number of elements pack up horizontally to automatically fill an equal, fluid width of their parent.

```javascript
import { Pack, PackItem } from "@staccx/bento"
```

## Standard Pack and PackItem
This is the standard Pack and PackItem.

<Preview>
  <Story name="Standard">
   <Pack justify="flex-start">
        <PackItem>Loan</PackItem>
        <PackItem>Credit</PackItem>
        <PackItem>Saving</PackItem>
        <PackItem>About this bank</PackItem>
    </Pack>
  </Story>
</Preview>

## NoWrap Pack and PackItem
This is the Pack with noWrap and PackItem.

<Preview>
  <Story name="NoWrap">
   <Pack middle noWrap>
        <PackItem>Loan</PackItem>
        <PackItem>Credit</PackItem>
        <PackItem>Saving</PackItem>
        <PackItem>About this bank</PackItem>
    </Pack>
  </Story>
</Preview>

## Props for Pack
<Props of={Pack} />

## Props for PackItem
<Props of={PackItem} />
