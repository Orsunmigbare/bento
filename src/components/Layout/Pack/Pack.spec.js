import React from "react"
import renderer from "react-test-renderer"
import { ThemeProvider } from "styled-components"
import baseTheme from "../../../theming/themes/baseTheme"
import Pack from "./Pack"
import PackItem from "./PackItem"

describe("Pack", () => {
  describe("Snapshots", () => {
    it("No props", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Pack>
              <PackItem>Loan</PackItem>
              <PackItem>Credit</PackItem>
              <PackItem>Saving</PackItem>
              <PackItem>About this bank</PackItem>
            </Pack>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("className", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Pack className="className">
              <PackItem>Loan</PackItem>
              <PackItem>Credit</PackItem>
              <PackItem>Saving</PackItem>
              <PackItem>About this bank</PackItem>
            </Pack>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("bottom", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Pack bottom>
              <PackItem>Loan</PackItem>
              <PackItem>Credit</PackItem>
              <PackItem>Saving</PackItem>
              <PackItem>About this bank</PackItem>
            </Pack>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("middle", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Pack middle>
              <PackItem>Loan</PackItem>
              <PackItem>Credit</PackItem>
              <PackItem>Saving</PackItem>
              <PackItem>About this bank</PackItem>
            </Pack>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("noWrap", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Pack noWrap>
              <PackItem>Loan</PackItem>
              <PackItem>Credit</PackItem>
              <PackItem>Saving</PackItem>
              <PackItem>About this bank</PackItem>
            </Pack>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("justify", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Pack justify="flex-start">
              <PackItem className="test">Loan</PackItem>
              <PackItem className="test">Credit</PackItem>
              <PackItem className="test">Saving</PackItem>
              <PackItem className="test">About this bank</PackItem>
            </Pack>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("Middle&Bottom", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Pack middle bottom>
              <PackItem className="test">Loan</PackItem>
              <PackItem className="test">Credit</PackItem>
              <PackItem className="test">Saving</PackItem>
              <PackItem className="test">About this bank</PackItem>
            </Pack>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("no Children", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Pack />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
