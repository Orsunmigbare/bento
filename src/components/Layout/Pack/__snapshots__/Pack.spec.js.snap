// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pack Snapshots Middle&Bottom 1`] = `
<ul
  className="sc-AykKC gmVKvP"
>
  <li
    className="sc-AykKD jxXobp test"
  >
    Loan
  </li>
  <li
    className="sc-AykKD jxXobp test"
  >
    Credit
  </li>
  <li
    className="sc-AykKD jxXobp test"
  >
    Saving
  </li>
  <li
    className="sc-AykKD jxXobp test"
  >
    About this bank
  </li>
</ul>
`;

exports[`Pack Snapshots No props 1`] = `
<ul
  className="sc-AykKC gmVKvP"
>
  <li
    className="sc-AykKD jxXobp"
  >
    Loan
  </li>
  <li
    className="sc-AykKD jxXobp"
  >
    Credit
  </li>
  <li
    className="sc-AykKD jxXobp"
  >
    Saving
  </li>
  <li
    className="sc-AykKD jxXobp"
  >
    About this bank
  </li>
</ul>
`;

exports[`Pack Snapshots bottom 1`] = `
<ul
  className="sc-AykKC jjfWXS"
>
  <li
    className="sc-AykKD jxXobp"
  >
    Loan
  </li>
  <li
    className="sc-AykKD jxXobp"
  >
    Credit
  </li>
  <li
    className="sc-AykKD jxXobp"
  >
    Saving
  </li>
  <li
    className="sc-AykKD jxXobp"
  >
    About this bank
  </li>
</ul>
`;

exports[`Pack Snapshots className 1`] = `
<ul
  className="sc-AykKC gmVKvP className"
>
  <li
    className="sc-AykKD jxXobp"
  >
    Loan
  </li>
  <li
    className="sc-AykKD jxXobp"
  >
    Credit
  </li>
  <li
    className="sc-AykKD jxXobp"
  >
    Saving
  </li>
  <li
    className="sc-AykKD jxXobp"
  >
    About this bank
  </li>
</ul>
`;

exports[`Pack Snapshots justify 1`] = `
<ul
  className="sc-AykKC hhkfZi"
>
  <li
    className="sc-AykKD jxXobp test"
  >
    Loan
  </li>
  <li
    className="sc-AykKD jxXobp test"
  >
    Credit
  </li>
  <li
    className="sc-AykKD jxXobp test"
  >
    Saving
  </li>
  <li
    className="sc-AykKD jxXobp test"
  >
    About this bank
  </li>
</ul>
`;

exports[`Pack Snapshots middle 1`] = `
<ul
  className="sc-AykKC dliJQw"
>
  <li
    className="sc-AykKD jxXobp"
  >
    Loan
  </li>
  <li
    className="sc-AykKD jxXobp"
  >
    Credit
  </li>
  <li
    className="sc-AykKD jxXobp"
  >
    Saving
  </li>
  <li
    className="sc-AykKD jxXobp"
  >
    About this bank
  </li>
</ul>
`;

exports[`Pack Snapshots no Children 1`] = `
<ul
  className="sc-AykKC gmVKvP"
/>
`;

exports[`Pack Snapshots noWrap 1`] = `
<ul
  className="sc-AykKC hCvsud"
>
  <li
    className="sc-AykKD jxXobp"
  >
    Loan
  </li>
  <li
    className="sc-AykKD jxXobp"
  >
    Credit
  </li>
  <li
    className="sc-AykKD jxXobp"
  >
    Saving
  </li>
  <li
    className="sc-AykKD jxXobp"
  >
    About this bank
  </li>
</ul>
`;
